libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+        if (ftp.replies->peek(reply) && (reply.chan == chan)) {
+                if ((i > 0) && comm_get_txspace(chan) < (2 * (packet_overhead() + MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL_LEN))) {
+                    // if this isn't the first packet we have to leave deadspace for the next message
+                    break;
+                }
+                if (!HAVE_PAYLOAD_SPACE(chan, FILE_TRANSFER_PROTOCOL)) {
+                    return;
+                }
+                ((uint16_t *)payload)[0] = reply.seq_number;
+                payload[2] = reply.session;
+                payload[3] = static_cast<uint8_t>(reply.opcode);
+                payload[4] = reply.size;
+                payload[5] = static_cast<uint8_t>(reply.req_opcode);
+                payload[6] = reply.burst_complete ? 1 : 0;
+                *(uint32_t *)(&payload[8]) = reply.offset;
+                memcpy(&payload[12], reply.data, sizeof(reply.data));
+                mavlink_msg_file_transfer_protocol_send(
+                    reply.chan,
+                    0, reply.sysid, reply.compid,
+                    payload);
+                ftp.replies->pop(reply);
+        } else {
+            break;
+        }
+    }
+}
+
+void GCS_MAVLINK::ftp_error(struct pending_ftp &response, FTP_ERROR error) {
+    response.opcode = FTP_OP::Nack;
+    response.data[0] = static_cast<uint8_t>(error);
+    response.size = 1;
+
+    // FIXME: errno's are not thread-local as they should be on ChibiOS
+    if (error == FTP_ERROR::FailErrno) {
+        // translate the errno's that we have useful messages for
+        switch (errno) {
+            case EEXIST:
+                response.data[0] = static_cast<uint8_t>(FTP_ERROR::FileExists);
+                break;
+            case ENOENT:
+                response.data[0] = static_cast<uint8_t>(FTP_ERROR::FileNotFound);
+                break;
+            default:
+                response.data[1] = static_cast<uint8_t>(errno);
+                response.size = 2;
+                break;
+        }
+    }
+}
+
+void GCS_MAVLINK::ftp_worker(void) {
+    pending_ftp request;
+    pending_ftp reply = {};
+    reply.session = -1; // flag the reply as invalid for any reuse
+
+    while (true) {
+        while (!ftp.requests->pop(request)) {
+            // nothing to handle delay ourselves a bit then check again, ideally we'd use conditional waits here
+            hal.scheduler->delay(10);
+        }
+
+        // if it's a rerequest and we still have the last response then send it
+        if ((request.sysid == reply.sysid) && (request.compid = reply.compid) &&
+            (request.session == reply.session) && (request.seq_number + 1 == reply.seq_number)) {
+            while (!ftp.replies->push(reply)) {
+                hal.scheduler->delay(10);
+            }
+            continue;
+        }
+
+        // setup the response
+        memset(&reply, 0, sizeof(pending_ftp));
+        reply.req_opcode = request.opcode;
```

Comment:
Again, brace initialisation would be nice.

Reply:
Not without designated initializers, which aren't available in C++11.